src/darwin/Framework/CHIP/ServerEndpoint/MTRServerEndpoint.h

```diff
@@ -0,0 +1,85 @@
+/**
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+#import <Matter/MTRAccessGrant.h>
+#import <Matter/MTRDefines.h>
+#import <Matter/MTRDeviceType.h>
+#import <Matter/MTRServerCluster.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * A representation of an endpoint implemented by an MTRDeviceController.
+ */
+MTR_NEWLY_AVAILABLE
+@interface MTRServerEndpoint : NSObject <NSCopying>
+
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+
+/**
+ * The provided endpointID must be in the range 1-65535.  The list of device
+ * types provided must be nonempty (but may include vendor-specific device
+ * types).
+ */
+- (nullable instancetype)initWithEndpointID:(NSNumber *)endpointID deviceTypes:(NSArray<MTRDeviceType *> *)deviceTypes;
+
+/**
+ * Add an access grant to the endpoint.
+ */
+- (BOOL)addAccessGrant:(MTRAccessGrant *)accessGrant;
+
+/**
+ * Remove an access grant from the endpoint.
+ */
+- (BOOL)removeAccessGrant:(MTRAccessGrant *)accessGrant;
+
+/**
+ * Add a server cluster to the endpoint.  This can only be done before the
+ * endpoint has been added to a controller.
+ *
+ * The cluster must not have the same cluster ID as another cluster on
+ * this endpoint.
+ *
+ * The cluster must not already be added to another endpoint.
+ */
+- (BOOL)addServerCluster:(MTRServerCluster *)serverCluster;
+
+@property (nonatomic, copy, readonly) NSNumber * endpointID;
+
+@property (nonatomic, copy, readonly) NSArray<MTRDeviceType *> * deviceTypes;
+
+/**
+ * The of entities that are allowed to access all clusters on this endpoint.
+ * If more fine-grained access control is desired, access grants should be
+ * defined on individual clusters.
+ *
+ * Defaults to empty set, which means no access granted.
+ */
+@property (nonatomic, copy, readonly) NSSet<MTRAccessGrant *> * accessGrants;
```

Comment:
What's the reason for using an NSSet here?

Reply:
There's no order of access grants, and I did not want to worry about what happens when you add a grant, then add the same grant, then remove the grant and what that should mean....  Making it a set makes it clear what the situation is, I thought.  Is there a reason to not use it?

Reply:
Per discussion: Use NSArray in the API, document the "set-like" behavior.

Reply:
Done.